staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -322,101 +322,103 @@ func unescapeSingleQuote(s string) (string, error) {
+			for i := 0; i < len(data); i++ {
+				// If in a single quoted string, look for the end of string
+				// ignoring delimiters.
+				if data[0] == '\'' {
+					if i > 0 && data[i] == '\'' && data[i-1] != '\\' {
+						// Return quoted string
+						return i + 1, data[:i+1], nil
 					}
-					fieldName = newS
-					if schema.AdditionalProperties != nil {
-						validFieldPath = validFieldPath.Key(fieldName)
+					continue
+				}
+				switch data[i] {
+				case '.', '[', ']': // delimiters
+					if i == 0 {
+						// Return the delimiter.
+						return 1, data[:1], nil
 					} else {
-						validFieldPath = validFieldPath.Child(fieldName)
+						// Return identifier leading up to the delimiter.
+						// The next call to split will return the delimiter.
+						return i, data[:i], nil
 					}
-				} else {
-					return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, "unexpected format of fieldName: "+fieldName)
-				}
-			} else if tok != scanner.Ident {
-				return pathOfFieldPath, invalidFieldError
-			} else {
-				if schema.AdditionalProperties != nil {
-					validFieldPath = validFieldPath.Key(fieldName)
-				} else {
-					validFieldPath = validFieldPath.Child(fieldName)
 				}
 			}
-
-			if schema.Properties != nil {
-				propertySchema, ok := schema.Properties[fieldName]
-				if ok {
-					schema = &propertySchema
-				} else {
-					return pathOfFieldPath, invalidFieldError
-				}
-			} else if schema.AdditionalProperties != nil {
-				schema = schema.AdditionalProperties.Structural
-			} else {
-				return pathOfFieldPath, invalidFieldError
+			if atEOF {
+				// Return the string.
+				return len(data), data, nil
 			}
+		}
+		return 0, nil, nil
+	})
 
-			if isMapSyntax {
-				if tok == scanner.EOF {
-					return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, "unexpected end of JSON path")
-				}
-				s.Scan()
-				if s.TokenText() != "]" {
-					return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, "expect ] but get: "+s.TokenText())
-				}
+	var tok string
+	for scanner.Scan() {
+		tok = scanner.Text()
+		switch tok {
+		case "[":
+			if !scanner.Scan() {
+				return nil, fmt.Errorf("unexpected end of JSON path")
+			}
+			tok = scanner.Text()
+			if len(tok) < 2 || tok[0] != '\'' || tok[len(tok)-1] != '\'' {
+				return nil, fmt.Errorf("expected single quoted string but got %s", tok)
+			}
+			unescaped, err := unescapeSingleQuote(tok[1 : len(tok)-1])
+			if err != nil {
+				return nil, fmt.Errorf("invalid string literal: %v", err)
+			}
+			if err := appendToPath(unescaped); err != nil {
+				return nil, err
+			}
+			if !scanner.Scan() {
+				return nil, fmt.Errorf("unexpected end of JSON path")
+			}
+			tok = scanner.Text()
+			if tok != "]" {
+				return nil, fmt.Errorf("expected ] but got %s", tok)
+			}
+		case ".":
+			if !scanner.Scan() {
+				return nil, fmt.Errorf("unexpected end of JSON path")
+			}
+			tok = scanner.Text()
+			if err := appendToPath(tok); err != nil {
```

Comment:
this should indicate to appendToPath the passed name must be a named property, right?

Reply:
Strictly speaking JSON path supports both access approaches for both maps and objects.  That said, I'm  okay what you suggest if we want to be more strict (and try to guide people to the most readable form of JSON path)